Implement ROS interface for live spot-aria with shared coordinate system and visualization in rviz

KavitShah1998 opened the pull request.

zephirefaith commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        self._is_connected = True
+        rospy.loginfo("Connected to Aria")
+
+    def _setup_aria(self) -> Tuple[aria.Device, aria.DeviceClient]:
```
also docstring

KavitShah1998 replied:
All class variables defined here are already declared in init. So we are good.

Added doctring.

zephirefaith commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+import os
```
file-level docstring to explain all the frames we work with in ROS and assume available in this file.

KavitShah1998 replied:
Should that be somewhere in the readme?

KavitShah1998 replied:
moved to README.md and linked to both `episodic_memory_robotic_fetch.py` and `aria_live_streamer.py`

zephirefaith commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        self.object_detector._core_objects = object_labels
+        self.object_detector._meta_objects = meta_objects
+
+    def get_running_avg_ariaWorld_T_marker(
```
should this really be a `AriaLiveStreamer` class function? This seems more generic, no? Maybe move to utils?

KavitShah1998 replied:
I also wanted to do so but in this case this function is also maintaining & updating member lists of positions array & quaternion array.

Not sure how to achieve this from an external function to a class. But if we achieve this, then it could be reused for SpotQRDetector too.

Let's discuss this.

KavitShah1998 replied:
As discussed, this should be moved to math_utils.py file.

KavitShah1998 replied:
moved

zephirefaith commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+@click.option("--do-update-iptables", is_flag=True, type=bool, default=False)
+@click.option("--read-only", is_flag=True, type=bool, default=False)
+@click.option("--debug", is_flag=True, type=bool, default=False)
+def main(do_update_iptables: bool, read_only: bool, debug: bool):
```
quick docstring to explain what main does.

KavitShah1998 replied:
added

zephirefaith commented on aria_data_loaders/aria_data_utils/aria_streamer.py:
```diff
@@ -692,13 +692,19 @@ def get_avg_spotWorld_T_marker_HAND(
-            if hand_mn_handcam_T_marker is not None:
+            hand_mn_handcam_T_marker = None
+            if hand_sp_handcam_T_marker is not None:
                 print("Trackedddd")
```
rospy.debug

KavitShah1998 replied:
Removed the print

zephirefaith commented on aria_data_loaders/aria_data_utils/conversions.py:
```diff
@@ -0,0 +1,241 @@
+from typing import Any, Dict, List, Optional, Tuple
```
renaming following convention we set up in previous PR #106

KavitShah1998 replied:
addressed

zephirefaith commented on aria_data_loaders/aria_data_utils/detector_wrappers/hod_wrapper.py:
```diff
@@ -0,0 +1,53 @@
+from hand_object_detector.faster_rcnn.faster_rcnn import _fasterRCNN
+
+
+class ObjectInHandDetectorWrapper(GenericDetector):
```
This needs basic testing before we merge it in.

zephirefaith replied:
Maybe remove this and only test+merge `object_in_hand_detector` in this PR.

KavitShah1998 replied:
Let's remove hand object detector from this PR & have a separate PR for this change?

KavitShah1998 replied:
moved to https://github.com/facebookresearch/spot-sim2real/pull/114

zephirefaith commented on aria_data_loaders/aria_data_utils/detector_wrappers/object_detector.py:
```diff
@@ -149,7 +168,31 @@ def _aria_fetch_demo_heuristics(
         return valid, score, stop
 
-    def process_frame(self, img_frame: np.ndarray):
+    def process_frame_online(
```
see if we can merge `process_frame_online` with `process_frame_offline`

KavitShah1998 replied:
It would be nice to merge the 2.

The "stop" condition is really valuable in on-line case as well as that will reduce compute at run-time

KavitShah1998 replied:
Rename process_frame_online to process_frame
Rename process_frame_offline to process_frame_with_stop

KavitShah1998 replied:
addressed

zephirefaith commented on aria_data_loaders/aria_data_utils/perception/object_in_hand_detector.py:
Needs basic testing.

KavitShah1998 replied:
moved to different PR https://github.com/facebookresearch/spot-sim2real/pull/114

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+FILTER_DIST = 2.4  # in meters (distance for valid detection)
+
+
+class SpotQRDetector:
```
Is `SpotQRDetector` defined in multiple places? Why is this another unique class? Does this entail we create `Spot<BespokeDetector>` class for every functionality?

Let's see if we can simplify this. Add detectors to Spot just like loggers accept handlers.

KavitShah1998 replied:
For now, I have moved this class to a new file `bd_spot_wrapper/spot_wapper/spot_qr_detector.py`

We can simplify it further in a separate effort as there is an issue already created for this task - https://github.com/facebookresearch/spot-sim2real/issues/90

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+        quat = se3_pose.rotation.normalize()
+        return sp.SE3(quat.to_matrix(), pos)
+
+    def _get_body_T_headcam(self, frame_tree_snapshot_head):
+        raise RuntimeError("This method is not used anymore. Please use get_spot_a_T_b")
+
+    def _get_spotWorld_T_handcam(
+        self, frame_tree_snapshot_hand, spot_frame: str = "vision"
+    ):
+        if spot_frame != "vision" and spot_frame != "odom":
+            raise ValueError("spot_frame should be either vision or odom")
+        spot_world_frame = spot_frame
+
+        hand_bd_wrist_T_handcam_dict = (
+            frame_tree_snapshot_hand.child_to_parent_edge_map.get(
+                "hand_color_image_sensor"
+            ).parent_tform_child
+        )
```
All of this needs to be removed as discussed previously.

KavitShah1998 replied:
This should be handled in its own PR. That will be a part of https://github.com/facebookresearch/spot-sim2real/issues/90

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+    ):
+        raise RuntimeError("This method is not used anymore. Please use get_spot_a_T_b")
+
+    def get_avg_spotWorld_T_marker_HAND(
```
Duplication of averaging code. Let's turn this into a utility and use in a generic manner.

KavitShah1998 replied:
We have a generic function for averaging. This will be taken care of in the PR to the issue - https://github.com/facebookresearch/spot-sim2real/issues/90

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+
+            # Spot - spotWorld_T_handcam computation
+            frame_tree_snapshot_hand = img_response_hand.shot.transforms_snapshot
+            hand_mn_body_T_handcam = self._get_body_T_handcam(frame_tree_snapshot_hand)
+            hand_mn_spotWorld_T_handcam = self._get_spotWorld_T_handcam(
+                frame_tree_snapshot_hand, spot_frame=spot_frame
```
utilize the generic function to get these transforms

KavitShah1998 replied:
This is already a part of issue https://github.com/facebookresearch/spot-sim2real/issues/90. Will be resolved in the PR to solve that issue

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+                    marker_position_from_dock_list.append(
+                        np.array(hand_mn_spotWorld_T_marker.translation)
+                    )
+                    marker_quaternion_form_dock_list.append(
```
typo: `marker_quaternion_[form->from]_dock_list`

KavitShah1998 replied:
addressed. thanks for finding this

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+
+class SpotAriaP1Node:
+    """
+    TODO: Add docstring
```
address todos

KavitShah1998 replied:
added

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+        # Instantiate static transform broadcaster for publishing marker w.r.t spotWorld transforms
+        self.static_tf_broadcaster = StaticTransformBroadcaster()
+
+        # Instantiate publisher for publishing goto pose for pick and place
+        self._goto_pose_pub_for_pick = rospy.Publisher(
+            "/goto_pose_for_pick", PoseStamped, queue_size=10
+        )
+        self._goto_pose_pub_for_place = rospy.Publisher(
+            "/goto_pose_for_place", PoseStamped, queue_size=10
+        )
+        self.goto_nav_pose_for_pick = None
+        self.goto_nav_pose_for_place = None
+
```
renaming as discussed previously, variables should not have verbs like "goto" in their names

KavitShah1998 replied:
addressed

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+                    f"Waiting for {self.spotWorld_T_aria} and {self.spotWorld_T_aria_pose_of_interest}"
+                )
+
+    def demo_trigger_callback(self, msg):
```
rename `demo_trigger` to `spot_fetch_and_handoff_trigger`, similar renaming for the flag

KavitShah1998 replied:
updated

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+        TODO: Add docstring
+        aria_pose is in ARIA CPF frame
+        """
+        # TODO: Device this logic
+        # return (2.079572916030884, 2.289151906967163, 0.7599455118179321)
```
clean-up

KavitShah1998 replied:
addressed

zephirefaith commented on spot_rl_experiments/configs/ros_topic_names.yaml:
```diff
@@ -14,6 +14,7 @@ IMAGE_SCALE: "/image_scale"
 ROBOT_STATE: "/robot_state"
 TEXT_TO_SPEECH: "/text_to_speech"
 
+ODOM_TOPIC: "/spot_odom"
```
@jimmytyyang @KavitShah1998 Does this new topic name follow Spot naming conventions?

KavitShah1998 replied:
I don't think we do have any naming convention for spot's ros frames. It would be a good idea to add one.

KavitShah1998 replied:
Let's rename to `/robot_odom`

KavitShah1998 replied:
addressed

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+    def _get_spotWorld_T_handcam(
+        self, frame_tree_snapshot_hand, spot_frame: str = "vision"
+    ):
+        if spot_frame != "vision" and spot_frame != "odom":
```
Instead of "vision" or "odom" these should probably be constants being accessed as part of Spot class?

KavitShah1998 replied:
added

zephirefaith commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,591 @@
+        camera_T_marker is used to compute spot_T_marker[i] and thus spotWorld_T_marker[i].
+        Then we average all spotWorld_T-marker to find average marker pose wrt spotWorld.
+        """
+        if spot_frame != "vision" and spot_frame != "odom":
```
Instead of "vision" or "odom" these should probably be constants being accessed as part of Spot class?

KavitShah1998 replied:
added them as frame names inside ros_frames.yaml file

KavitShah1998 replied:
@zephirefaith let's add a nested name structure inside ros_frame_names.yaml where we can describe all generic frames and specific frames too .. Eg:

```
rf.SPOT.WORLD_ODOM
rf.SPOT.BASE
rf.SPOT.ARM.WRIST_0
rf.ARIA.CPF
rf.ARIA.DEVICE
rf.ARIA_WORLD
rf,.MARKER
...
```

zephirefaith commented on bd_spot_wrapper/spot_wrapper/spot.py:
```diff
@@ -863,6 +869,16 @@ def convert_transformation_from_BD_to_magnum(self, bd_transformation_dict: dict)
         mn_transformation = mn.Matrix4.from_(rotation_matrix, translation)
         return mn_transformation
 
+    def get_vision_tform_body_TransformStamped(self, frame_tree_snapshot):
```
this will get refactored when function names are changed

KavitShah1998 replied:
addressed

zephirefaith commented on spot_rl_experiments/spot_rl/envs/skill_manager.py:
```diff
@@ -68,15 +68,18 @@ class SpotSkillManager:
 
         # Power on the robot
-        self.spot.power_robot()
+        if should_stand_on_start:
+            self.spot.power_robot()
+        else:
+            self.spot.power_on()
```
@KavitShah1998 add comment here explaining what is this and why

KavitShah1998 replied:
Discussed offline, we remove it

KavitShah1998 replied:
removed

zephirefaith commented on spot_rl_experiments/spot_rl/launch/local_only.sh:
We should have some documentation somewhere describing each ROS node we spawn

KavitShah1998 replied:
https://github.com/facebookresearch/spot-sim2real/issues/116

zephirefaith commented on spot_rl_experiments/spot_rl/utils/helper_nodes.py:
```diff
@@ -59,6 +72,21 @@ def publish_msgs(self):
+            pose = self.spot.get_vision_T_body_Pose(robot_kinematic_snapshot_tree)
+            msg.pose.pose = pose
+            msg.header.stamp = rospy.Time.now()
+            msg.child_frame_id = "spot"
+            msg.header.frame_id = "spotWorld"
```
shouldn't these be constants in Spot class rather than hardcoded strings here?

KavitShah1998 replied:
From ros_topics.yaml

KavitShah1998 replied:
addressed

zephirefaith commented on spot_rl_experiments/spot_rl/utils/robot_subscriber.py:
```diff
@@ -36,6 +36,7 @@ class SpotRobotSubscriberMixin:
 
     def __init__(self, spot=None, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        rospy.loginfo(f"name - {self.node_name}")
```
what does this class do? Can we get a class docstring?

KavitShah1998 replied:
Actually this was a debug print. I will remove this.

KavitShah1998 replied:
removed

KavitShah1998 commented on .gitmodules:
```diff
@@ -13,3 +13,6 @@
 	url = https://github.com/naokiyokoyama/mask_rcnn_detectron2
 	branch = main
 	ignore = dirty
+[submodule "third_party/hand_object_detector"]
+	path = third_party/hand_object_detector
+	url = https://github.com/ddshan/hand_object_detector
```
@zephirefaith  do we need to add hand-obj detector right now?
I suggest let's remove this as we are not using it?

KavitShah1998 replied:
removed and moved to a separate PR - https://github.com/facebookresearch/spot-sim2real/pull/114

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        self._in_index = 0
+        self._out_index = 0
+        self._is_connected = False
+        self.aria_pose_sub = rospy.Subscriber(
```
How about `aria_open_loop_pose_sub` ?
As it it reading for pose not trajectory?

KavitShah1998 replied:
updated.

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        )  # GOD KNOW WHY THIS IS NEEDED -> https://github.com/clovaai/CRAFT-pytorch/issues/84#issuecomment-574683857
+        return img
+
+    def on_image_received(self, image: np.ndarray, record: ImageDataRecord):
```
Actually this is an un-used method. Removed

KavitShah1998 replied:
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+            rospy.logdebug("Received image")
+            self.aria_rgb_frame = image
+
+    def on_pose_received(self, msg: PoseStamped):
```
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+            self.aria_rgb_frame = image
+
+    def on_pose_received(self, msg: PoseStamped):
+        if self._is_connected:
+            rospy.logdebug("Received pose")
+            self.aria_pose_device = ros_pose_to_sophus(msg.pose)
+            self.aria_pose = self.aria_pose_device * self._device_T_cpf
+            self.aria_ros_pose = sophus_to_ros_pose(self.aria_pose)
+            odom_msg = Odometry()
+            odom_msg.header.frame_id = "ariaWorld"
+            odom_msg.child_frame_id = "ariaDevice"
+            odom_msg.header.stamp = rospy.Time.now()
+            odom_msg.pose.pose = self.aria_ros_pose
+            self.aria_odom_publisher.publish(odom_msg)
+            msg.pose = self.aria_ros_pose
+            self.aria_cpf_publisher.publish(msg)
```
Good practice to add the `if` condition check here. But just curious, will we receive any new poses if device got disconnected for whatever reason?

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+
+    def publish_pose_of_interest(self, pose: Pose, marker_scale: float = 0.1):
+        """
+        Publishes current pose of Aria as a pose of interest for Spot
```
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        self._out_index += 1
+
+    def get_frame(self) -> Optional[Dict[str, Any]]:
+        """
```
+1

KavitShah1998 replied:
As discussed offline, TL;DR - Move logic of update_frame inside get_frame, rename get_frame to get_latest_frame, delete member variable self._latest_frame() as it is currently not used anywhere.

KavitShah1998 replied:
addressed

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        self.update_frame()
+        return self._latest_frame
+
+    def update_frame(self):
```
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+        detect_hand_object: bool = False,
+        object_label="milk bottle",
+    ) -> Tuple[np.ndarray, dict]:
+        """
+        Process the frame
+        """
```
added a docstring similar to aria_streamer.py

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,469 @@
+
+    def initialize_object_detector(self, outputs: dict = {}, object_labels: list = []):
+        """
+        Initialize the object in hand detector
```
updated

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_sdk_utils.py:
```diff
@@ -0,0 +1,73 @@
+    subprocess.run(update_iptables_cmd)
+
+
+@contextmanager
+def ctrl_c_handler(signal_handler=None):
+    class ctrl_c_state:
+        def __init__(self):
+            self._caught_ctrl_c = False
+
+        def __bool__(self):
+            return self._caught_ctrl_c
+
+    state = ctrl_c_state()
+
+    def _handler(sig, frame):
+        state._caught_ctrl_c = True
+        if signal_handler:
+            signal_handler()
+
+    original_sigint_handler = signal.getsignal(signal.SIGINT)
+    signal.signal(signal.SIGINT, _handler)
+
+    try:
+        yield state
+    finally:
+        signal.signal(signal.SIGINT, original_sigint_handler)
+
+
+def quit_keypress():
+    key = cv2.waitKey(1)
+    # Press ESC, 'q'
+    return key == 27 or key == ord("q")
```
This function is also not being used anywhere. Removing it.

KavitShah1998 commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,336 @@
+
+    def __init__(self, spot: Spot, verbose=False):
+        # Get Spot Skill Manager
+        self.skill_manager = SpotSkillManager(
+            should_stand_on_start=False,
+            use_mobile_pick=True,
+        )  # @FIXME: NEED THIS HERE AS IT INITIALIZES A Ros-node which we need here after
```
Discuss with @zephirefaith

KavitShah1998 replied:
Discussed and added a FIXME and updated in an issue - https://github.com/facebookresearch/spot-sim2real/issues/115

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_sdk_utils.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@contextmanager
+def ctrl_c_handler(signal_handler=None):
+    class ctrl_c_state:
+        def __init__(self):
+            self._caught_ctrl_c = False
+
+        def __bool__(self):
+            return self._caught_ctrl_c
+
+    state = ctrl_c_state()
+
+    def _handler(sig, frame):
+        state._caught_ctrl_c = True
+        if signal_handler:
+            signal_handler()
+
+    original_sigint_handler = signal.getsignal(signal.SIGINT)
+    signal.signal(signal.SIGINT, _handler)
+
+    try:
+        yield state
+    finally:
+        signal.signal(signal.SIGINT, original_sigint_handler)
```
Remove this

KavitShah1998 replied:
addressed

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_streamer.py:
```diff
@@ -1,25 +1,21 @@
-from spot_rl.models.owlvit import OwlVit
-from spot_wrapper.spot import Spot, SpotCamIds, image_response_to_cv2
+from spot_wrapper.spot import Spot
+from spot_wrapper.spot_qr_detector import SpotQRDetector
```
@jimmytyyang  is it okay to move SpotQRDetector code inside bd_spot_wrapper?

jimmytyyang replied:
yeah, I think it should be fine

KavitShah1998 commented on aria_data_loaders/aria_data_utils/conversions.py:
```diff
@@ -0,0 +1,304 @@
+    return sp.SE3(rotation, translation)
+
+
+def compute_avg_sophus_SE3_from_nplist(
+    a_T_b_position_list: List[np.ndarray], a_T_b_quaternion_list: List[np.ndarray]
+) -> Optional[sp.SE3]:
+    """
+    Computes the average transformation of aria world frame to marker frame
+
+    Args:
+        a_T_b_position_list (List[np.ndarray]): List of positions of marker frame in aria world frame
+        a_T_b_quaternion_list (List[np.ndarray]): List of quaternions of marker frame in aria world frame
+
+    Returns:
+        avg_sp_se3 (sp.SE3): Average of all the transformations as a Sophus SE3 object
+    """
+    assert len(a_T_b_position_list) == len(
+        a_T_b_quaternion_list
+    ), "Position and Quaternion lists must be equal length"
+
+    if len(a_T_b_position_list) == 0:
+        return None
+
+    a_T_b_position_np = np.array(a_T_b_position_list)
+    avg_a_T_b_position = np.mean(a_T_b_position_np, axis=0)
+
+    a_T_b_quaternion_np = np.array(a_T_b_quaternion_list)
+    avg_a_T_b_quaternion = np.mean(a_T_b_quaternion_np, axis=0)
+
+    avg_sp_se3 = sp.SE3(
+        Rotation.from_quat(avg_a_T_b_quaternion).as_matrix(), avg_a_T_b_position
+    )
+
+    return avg_sp_se3
```
Move to math_utils.py

KavitShah1998 replied:
updated.

KavitShah1998 commented on aria_data_loaders/aria_data_utils/detector_wrappers/generic_detector_interface.py:
```diff
@@ -11,5 +13,5 @@ def enable_detector(self):
         self.is_enabled = False
 
-    def process_frame(self, frame: np.ndarray):
+    def process_frame(self, frame: np.ndarray) -> Dict[str, Any]:
```
Modify the return type to also include image as np.ndarray

KavitShah1998 replied:
addressed

KavitShah1998 commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,336 @@
+    def __init__(self, spot: Spot, verbose=False):
+        # Get Spot Skill Manager
+        self.skill_manager = SpotSkillManager(
+            should_stand_on_start=False,
```
Remove this flag, add spot.nav & spot.sit logic here which brings robot next to dock in a way it sees the QR marker

KavitShah1998 replied:
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,336 @@
+        # column 3 from spotWorld_T_cpf represents cpf's z axis in spotWorld frame
+        cpf_z_axis_in_spotWorld = aria_pose.matrix()[:3, 2]
+
+        # Project cpf's z axis onto xy plane of spotWorld frame by ignoring z component
+        xy_plane_projection_array = np.array([1.0, 1.0, 0.0])
+        projected_cpf_z_axis_in_spotWorld_xy = np.multiply(
+            cpf_z_axis_in_spotWorld, xy_plane_projection_array
+        )
+        orientation = float(
+            np.arctan2(
+                projected_cpf_z_axis_in_spotWorld_xy[1],
+                projected_cpf_z_axis_in_spotWorld_xy[0],
+            )
+        )  # tan^-1(y/x)
```
```suggestion
        orientation = float(
            np.arctan2(
                cpf_z_axis_in_spotWorld[1],
                cpf_z_axis_in_spotWorld[0],
            )
        )  # tan^-1(y/x)
```

KavitShah1998 replied:
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,336 @@
+        x_component = cpf_z_axis_in_spotWorld[0]
+        y_component = cpf_z_axis_in_spotWorld[1]
+
+        # Project cpf's z axis onto xy plane of spotWorld frame by ignoring z component
+        xy_plane_projection_array = np.array([1.0, 1.0, 0.0])
+        projected_cpf_z_axis_in_spotWorld_xy = np.multiply(
+            cpf_z_axis_in_spotWorld, xy_plane_projection_array
+        )
+        rotation = float(
+            np.arctan2(
+                projected_cpf_z_axis_in_spotWorld_xy[1],
+                projected_cpf_z_axis_in_spotWorld_xy[0],
+            )
+        )  # tan^-1(y/x)
```
```suggestion
        x_component = cpf_z_axis_in_spotWorld[0]
        y_component = cpf_z_axis_in_spotWorld[1]
        orientation = float(
            np.arctan2(
                y_component,
                x_component,
            )
        )  # tan^-1(y/x)
```

KavitShah1998 replied:
addressed

KavitShah1998 commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,336 @@
+        nav_for_place: Tuple[float, float, float],
+        place_loc: Tuple[float, float, float],
+    ):
+        """ """
```
Add docstring

KavitShah1998 replied:
added

KavitShah1998 commented on aria_data_loaders/aria_data_utils/spot_aria_p1_node.py:
```diff
@@ -0,0 +1,336 @@
+
+
+@click.command()
+@click.option("--data-path", help="Path to the data directory", type=str)
+@click.option("--vrs-name", help="Name of the vrs file", type=str)
+@click.option("--dry-run", type=bool, default=False)
+@click.option("--verbose", type=bool, default=True)
+@click.option("--use-spot/--no-spot", default=True)
+@click.option("--object-names", type=str, multiple=True, default=["smartphone"])
+@click.option("--qr/--no-qr", default=True)
```
Remove unwanted flags

KavitShah1998 replied:
removed

KavitShah1998 commented on spot_rl_experiments/spot_rl/envs/place_env.py:
```diff
@@ -276,7 +279,7 @@ def get_observations(self):
     with spot.get_lease(hijack=True):
         spot.power_robot()
-        place_controller = PlaceController(config, spot, use_policies=True)
+        place_controller = PlaceController(config, spot, use_policies=False)
```
Make `use_policies` this an argument

KavitShah1998 replied:
addressed

KavitShah1998 commented on spot_rl_experiments/spot_rl/launch/local_only.sh:
```diff
@@ -3,13 +3,17 @@ tmux kill-session -t roscore
 tmux new -s propio_pub -d '$CONDA_PREFIX/bin/python -m spot_rl.utils.helper_nodes --proprioception'
 tmux new -s tts_sub -d '$CONDA_PREFIX/bin/python -m spot_rl.utils.helper_nodes --text-to-speech'
+tmux new -s spotWorld_static_tf2_pub -d 'rosrun tf2_ros static_transform_publisher 0 0 0 0 0 0 /map /spotWorld'
+# tmux new -s rviz -d 'rosrun rviz rviz -d $CONDA_PREFIX/envs/spot_rl/lib/python3.6/site-packages/spot_rl/utils/spot_world.rviz'
```
Update readme and add instruction to run it separately.

KavitShah1998 replied:
Updated.

KavitShah1998 commented on spot_rl_experiments/spot_rl/models/owlvit.py:
```diff
@@ -11,24 +11,46 @@
-        self.model = OwlViTForObjectDetection.from_pretrained(
-            "google/owlvit-base-patch32"
-        )
+        if version == 1:
+            self.model = OwlViTForObjectDetection.from_pretrained(
+                "google/owlvit-base-patch32"
+            )
+            self.processor = OwlViTProcessor.from_pretrained(
+                "google/owlvit-base-patch32"
+            )
+        else:
+            self.processor = Owlv2Processor.from_pretrained(
+                "google/owlv2-base-patch16-ensemble"
+            )
+            self.model = Owlv2ForObjectDetection.from_pretrained(
+                "google/owlv2-base-patch16-ensemble"
+            )
```
Make it a version protected `import`

KavitShah1998 replied:
addressed

KavitShah1998 commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,638 @@
+def main(do_update_iptables: bool, read_only: bool, debug: bool):
+    """
+    Main function initializes the AriaLiveReader object, its detectors and runs detectors on every frame while publishing
+    necessary data on different ROS topics
+
+    Args:
+        do_update_iptables (bool, optional): Whether to update iptables to allow for streaming. Defaults to False.
```
added

jimmytyyang commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,626 @@
+TF Frames used for Spot-Aria integration
+
+map: RVIZ map frame (necessary as this acts as root node in tf tree). Always exists with rviz
+spotWorld: Spot's world frame (defined by where spot's home is reset). The local_launsh shell script publishes a static tf of spotWorld wrt map
```
typo: launch

KavitShah1998 replied:
thanks for finding this. updated.

jimmytyyang commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,626 @@
+        self.marker_quaternion_list = (
+            []
+        )  # type: List[np.ndarray] # List of quaternions as np.ndarray (x, y, z, w)
+        self.fix_list_length = 10  # TODO: Add this to config
```
please fix this TODO. Thank you.

KavitShah1998 replied:
updated. currently hard set the value inside math_utils.py

jimmytyyang commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,626 @@
+        this_file_path = os.path.dirname(os.path.realpath(__file__))
+        self._device_T_cpf = np_matrix3x4_to_sophus_SE3(
+            np_matrix=np.load(
+                os.path.join(this_file_path, "../device_model/device_T_cpf.npy")
```
where do you get device_T_cpf.npy?

KavitShah1998 replied:
device_T_cpf transformation is a constant transform.
Device frame is the Left Slam Camera, CPF is a frame at center of aria.

More info on frames can be found here - https://facebookresearch.github.io/projectaria_tools/docs/data_formats/coordinate_convention/3d_coordinate_frame_convention

KavitShah1998 replied:
added the link to that line in code. @rutadesai

jimmytyyang commented on aria_data_loaders/aria_data_utils/episodic_memory_robotic_fetch.py:
```diff
@@ -0,0 +1,435 @@
+import logging
```
is this a main script to run the demo? if it is, then this should be move to some example folder

KavitShah1998 replied:
Yes @jimmytyyang  this is the main driver node on PC that runs spot.
I didn't understand why it should be moved to another folder though. Can you elaborate on that please?

akshararai replied:
I agree, this belongs in the utils/ folder.
However, we don't have a clean executables/ or examples/ folder.
We should make one and move this and the scripts for running Spot alone experiments there.

KavitShah1998 replied:
Issue link - https://github.com/facebookresearch/spot-sim2real/issues/125

jimmytyyang commented on spot_rl_experiments/spot_rl/launch/local_only.sh:
```diff
@@ -3,13 +3,15 @@ tmux kill-session -t roscore
 tmux new -s img_pub -d '$CONDA_PREFIX/bin/python -m spot_rl.utils.img_publishers --local'
 tmux new -s propio_pub -d '$CONDA_PREFIX/bin/python -m spot_rl.utils.helper_nodes --proprioception'
 tmux new -s tts_sub -d '$CONDA_PREFIX/bin/python -m spot_rl.utils.helper_nodes --text-to-speech'
+tmux new -s spotWorld_static_tf2_pub -d 'rosrun tf2_ros static_transform_publisher 0 0 0 0 0 0 /map /spotWorld'
```
do we need this to launch the script if we do not use Aria?

KavitShah1998 replied:
Yes, because now you can also visualize real-time location of spot inside rviz. Without needing aria.

For visualizing it in rviz, you will need this publisher.

zephirefaith replied:
@jimmytyyang Does the above make sense to you?

jimmytyyang commented on spot_rl_experiments/spot_rl/models/owlvit.py:
```diff
@@ -11,24 +11,64 @@
         self.device = (
             torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
         )
+        if version == 1:
```
Is the version one be the one we used before?

KavitShah1998 replied:
Yes, we were using Owlvit i.e. version1 before. Aria code uses OwlV2 or version2 but this may change again in near future.

jimmytyyang reviewed (commented):
Thank you so much for this PR. Overall, I think they look good! But one concern I have is that will this affect the existing skill execution? For example, for users who only want to do development without using Aria, will this PR introduce dependency to Aria installation? I can see some files import Aria stuff. Is it possible to install spot-sm2real only the skill part without installing Aria code for the ease of downstream usecases? I just want to ensure that this PR is well-tested in terms of installation and package dependency. Thank you!

rutadesai commented on spot_rl_experiments/spot_rl/envs/skill_manager.py:
```diff
@@ -119,7 +130,7 @@ def __initiate_controllers(self):
         self.gaze_controller = GazeController(config=self.pick_config, spot=self.spot)
         self.place_controller = PlaceController(
-            config=self.place_config, spot=self.spot, use_policies=False
+            config=self.place_config, spot=self.spot, use_policies=use_policies
```
unless it is described elsewhere in the code/readme, we may want to specify where these policies are located/coming from e.g., a specific config or path? If we are providing these policies, may be we can add in the readme a way to download them as well from some place just like we did for hab3.0 release.

zephirefaith replied:
@tusharsangam

tusharsangam replied:
@rutadesai config.yaml contains the path to individual policy weights. While setup instructions page describes how to download these weights [Like this one](https://github.com/facebookresearch/spot-sim2real/blob/main/installation/SETUP_INSTRUCTIONS.md#:~:text=%23%20Download%20weights%20(need%20only%20once)%0Agit%20clone%20https%3A//huggingface.co/spaces/jimmytyyang/spot%2Dsim2real%2Ddata%0Aunzip%20spot%2Dsim2real%2Ddata/weight/weights.zip%20%26%26%20rm%20%2Drf%20spot%2Dsim2real%2Ddata%20%26%26%20cd%20../)
However there is no individual section in readme that goes over each policy details like what are its observations, functions & output actions, etc. Is this the section you mean to add ?
Since I haven't worked on hab 3.0 can you please point me to the readme section you are referring to get more clear idea ?

KavitShah1998 replied:
As discussed offline with @rutadesai offline, this is not needed here as info about policies is provided in the SETUP_INSTRUCTIONS.md which use will go through at the start of setup process.

rutadesai reviewed (approved):
Added some comments -- mostly small nits. Looks good overall. Really appreciated all the docstrings and detailed comments and things organized nicely in their own classes and utils files. Great job! Let's get this landed soon :)

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+        /aria_current_pose: PoseStamped msg containing current pose of aria (cpf) w.r.t ariaWorld
+        /aria_pose_of_interest: PoseStamped msg containing pose of interest (cpf) w.r.t ariaWorld
+        /aria_odom: Odometry msg containing odometry info (in cpf) w.r.t ariaWorld
+        /aria_pose_of_interest_marker: Marker msg containing pose of interest (cpf) w.r.t ariaWorld
```
not sure what this comment means. Is it the pose of Aria when the marker was seen?

zephirefaith replied:
This topic publishes a Marker message so we may visualize `/aria_pose_of_interest` in rviz. We went with a ROS marker as we could make it whatever shape+color we wanted and disambiguate w/`/aria_current_pose` easily via code for debugging.

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+        """
+        self._is_connected = False
+        self.device.streaming_manager.streaming_client.unsubscribe()  # type: ignore
+        self.device_client.disconnect(self.device)  # type: ignore
```
Also add a message here ?
rospy.loginfo("Aria disconnected")

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+
+        return device, device_client
+
+    def _rotate_img(self, img: np.ndarray, num_of_rotation: int = 3) -> np.ndarray:
```
in general, a function like this can sit outside of a class, since it is not really specific to Aria. For example it doesn't use any of the objects, etc from the current class. This was it can be used elsewhere without needing to instantiate aria. Is there some version of utils.py that this can be moved to?

zephirefaith replied:
I want us to create a 3rd common package within Spot-sim2real which houses all the shared perception functionalities+utils. That way Spot or Aria can just minimally use that as dependency. The alternative is housing this utils in Aria or Spot, where one now has a heavy-weight dependency on the other complex package.

KavitShah1998 replied:
@akshararai @zephirefaith this should be a new PR.

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+
+        self._out_index += 1
+
+    def get_latest_frame(self) -> Optional[Dict[str, Any]]:
```
or get_latest_pose_and_image()

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+                avg_marker_T_ariaWorld = self.avg_ariaWorld_T_marker.inverse()
+
+                # Publish marker pose in ariaWorld frame
+                self.static_tf_broadcaster.sendTransform(
```
should this be published every time this function is called, or only the first time the QR is detected?

Basically, should `if camera_T_marker is not None:` be within the previous `if detect_qr:` loop?

zephirefaith replied:
We need to publish this per cycle otherwise the rviz marker will fade due to old timestamp.
That said, `if camera_T_marker is not None:` can be within previous condition-set as we do not have any alt way of getting `camera_T_marker`

KavitShah1998 replied:
As @zephirefaith  mentioned
1. The TF that we publish here gets registered to the TF tree which contains (spotWorld->spot->marker). This current TF publishes (marker->ariaWorld). Speaking about marker->ariaWorld, many-a-times, due to noise, ariaWorld might be shifted from where it actually is. If user keeps looking at marker for a few seconds steadily, this noise gets reduced and that (marker->ariaWorld) detections are more "acceptable". So with that idea, it is good and useful to keep publishing (marker->ariaWorld) transformation for making sure the transform is relatively noise-free.
2. It would make no difference if we move the logic from within `if camera_T_marker is not None:` to inside `if detect_qr:`. If detect_qr bool is false, it is just going to get by-passed all the time. Logic will never enter there so we should be good.

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+                img_metadata=None,
+            )
+
+        if detect_objects:
```
is it either detect_object or detect_qr? What happens if both are false?

zephirefaith replied:
We had both ON while running P1 demo as we wanted to detect QR when it pops up and also detect objects throughout the episode. If both are false, then function just returns the input image as is and an empty dictionary.
We coded it like this as it made easy to test one or the other while debugging.

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+        """
+        self.object_detector = ObjectDetectorWrapper()
+        self.object_detector.enable_detector()
+        meta_objects: List[str] = []
```
what is meta_objects?

zephirefaith replied:
In our OWL-ViT  based hand-object intersection logic, we ask OWL-ViT to detect a hand and the object of interest and then see if the bboxes intersect. "hand" then is the meta-object as it is something we provided, while core_object is the object_label passed by the user.

This was a noisy logic and we plan on replacing this with more sophisticated hand-object-interaction detectors.

rutadesai replied:
Let's add a comment about what meta-objects are for now?

KavitShah1998 replied:
added

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+        rospy.logwarn("Not updating iptables")
+    object_name = "bottle"
+
+    rate = rospy.Rate(5)
```
or a config parameter?

akshararai commented on aria_data_loaders/aria_data_utils/aria_live_streamer.py:
```diff
@@ -0,0 +1,598 @@
+import os
```
need copyright stuff

akshararai commented on aria_data_loaders/aria_data_utils/aria_streamer.py:
```diff
@@ -1,25 +1,22 @@
 import logging
```
copyright stuff?

akshararai commented on aria_data_loaders/aria_data_utils/conversions.py:
```diff
@@ -0,0 +1,269 @@
+
+###################################################################################################
+# Taken from:
+# https://github.com/facebookresearch/home-robot/blob/1f393f3a244af023684668539d075e7acabde774/src/home_robot/home_robot/utils/geometry/_base.py#L58
```
thanks for acknowledging :)

akshararai commented on aria_data_loaders/aria_data_utils/detector_wrappers/object_detector.py:
```diff
@@ -55,7 +55,10 @@ def _init_object_detector(
         self.object_detector = OwlVit(
-            [object_labels], score_threshold=0.125, show_img=self.verbose
+            [object_labels],
+            score_threshold=0.125,
```
config parameter?

akshararai reviewed (commented):
done until aria_data_loaders/aria_data_utils/episodic_memory_robotic_fetch.py
Need to stop, this PR is really tooooo long to be reviewed properly.

KavitShah1998 commented on README.md:
````diff
@@ -182,19 +182,110 @@ git checkout main
+    python -m streaming_subscribe
+    ```
+
+#### Step2. Run Off-the-shelf localization for Aria
````
@rutadesai currently we use the internal localization engine for Aria. We do not have a tested way for how an external person will run localization. My reason for putting it there was so people will know that there needs to be a step like that to be taken.

For internal users, we have a separate doc that explains the steps to be taken for running internal localization. @zephirefaith should we link that doc here?

zephirefaith replied:
@rutadesai @KavitShah1998 @akshararai @jimmytyyang Need some discussion here. Running SLAM on Aria is done through internal code which can not be linked here. One thing we could do is document all the expected outputs from SLAM running on Aria and move on. IIRC, I was told that open-loop trajectory will be made available in the CLI at some point so we can update this section to point to those public instructions.

KavitShah1998 replied:
Added a message here based on the discussion on slack.

KavitShah1998 commented on spot_rl_experiments/configs/ros_frame_names.yaml:
```diff
@@ -0,0 +1,19 @@
+
+ARIA_WORLD: "ariaWorld"
+ARIA: "aria" # aria in cpf frame
+ARIA_DEVICE: "ariaDevice" # aria in device frame
```
Its added in README.md

KavitShah1998 commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
+- spotWorld:  Spot's world frame (defined by where spot's home is reset). The local_launch shell script publishes a static tf of spotWorld wrt map
+- spot:       Spot's base frame (defined by current location of spot). This frame is published by Proprioception node in local_launch shell script
+- marker:     Marker frame (defined by the center of spot dock's april tag). This frame is published by Episodic_memory_robotic_fetch node at the start of the node when spot looks at the marker and detects it
+- ariaWorld:  Aria's world frame (defined by location where Aria's SLAM pipeline runs). This frame is published by AriaLiveReader node
```
added

KavitShah1998 commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
+### Run ARIA - Spot Fetch tasks from LIVE Aria Stream
+- This mode requires a Multi - PC setup.
+    - PC #1 can be any Linux workstation that is capable to run low-level spot skills
+    - PC #2 should be a linux laptop (with good compute) which **will run the live streamer code in your bagpack while connected to Aria device via USB**
```
removed.

KavitShah1998 commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
+- Spot will standup from dock & navigate to the left of the dock & sit down so as to view the marker.
+- After you see `spot`, `spotWorld` and `marker` frames in RVIZ, you need to get Aria in the same frame system.
+- Run the steps for PC#2 as described [below](#steps-for-pc-2) to enable Aria's streaming
+- Once Aria's streaming has started and you are wearing Aria while having the PC#2 (i.e. laptop) in your bagback, **move to exactly where the spot is sitting and view the QR code of spot's dock for 2 seconds**. Make sure you can also see `aria` & `ariaWorld` frames in the same RVIZ window and the odometry of aria constantly updating as you are moving.
```
@akshararai 
It doesn't need to be where spot is sitting. It should just be within 2m range of marker (as QR detection was accurate there).

User would not know where is the 2m radius around marker so it would be easier for user to go near where spot is ... because we make spot to go and sit at a hard-defined spot to view the marker. This is just to make things easy for user.

But I have removed this requirement from readme now.

zephirefaith commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
 - If you are done with demo of one of the above code and want to run another code, you do not need to re-run other sessions and nodes. Running a new command in the same terminal will work just fine. But **make sure to bring robot at home location and reset its home** using `spot_reset_home` in the same terminal
 
 ## :eyeglasses: Run Spot-Aria project code
```
Agree with the suggestion. We'll move the readme to aria_data_loaders folder (a name I do not like anymore, it is not accurate) and link the readme to this readme.

zephirefaith replied:
We are addressing all comments in this readme first, to keep things easy-to-find. Will move them to Aria ReadMe if details look good.

KavitShah1998 replied:
Thank you @zephirefaith for taking care of this one.

zephirefaith commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
+- Also make sure that the frames in the system are making sense w.r.t. real world configuration
+- Move in the environment and look at the object of interest from a good viewpoint from where spot can grasp it. You would see a green arrow appear at that location in RVIZ which showcases that system has locked on the object of interest at the shown pose.
+- Move in the environment and finally sit a location where there is navigable space in front of you. You would see a purple arrow appear at that location in RVIZ which showcases that system has registered your last location as place location at the shown pose.
+- Run [Step7](#step7-in-a-new-terminal-run-this-publisher-once-all-frames-are-registered-inside-rviz-and-spot-fetch-tasks-should-be-initialized) to initiate spot-fetch.
```
This is easy to implement and should be the de-facto method. I think we were using this protocol for testing and then due to time-constraints went with this for final P1 demo video too.

zephirefaith replied:
@akshararai What would be a reasonable automatic way of triggering this? Is Spot capable of listening to text? That could be one way. That said, can we merge this PR first and then make iterative changes in next version? I will add an issue for this.

KavitShah1998 replied:
@akshararai 
The "demo trigger" is used to start the actual Spot fetch tasks.

When the human is moving around in the environment, he sees / picks the object of interest, but that is not sufficient to trigger SpotFetch as human was not have moved yet.

Only after human has actually moved to their location can we finally start the Spot Fetch.

Until then, the execution of code logic DOES NOT have any way of knowing if spot fetch tasks should be triggered.


Alternative : We can try to integrate our Whisper code into this P1 demo stack, where we can use audio2text and make sure user has given a command like "Hey Spot please fetch the items to me" or something.

cc. @zephirefaith @rutadesai @jimmytyyang

zephirefaith commented on bd_spot_wrapper/spot_wrapper/spot.py:
```diff
@@ -24,7 +24,12 @@
 import quaternion
+import rospy
 import sophus as sp
+from aria_data_utils.conversions import (
```
Agreed. Check out my response about a 3rd common pkg in a previous comment.

KavitShah1998 replied:
Thanks @zephirefaith  for creating this issue - https://github.com/facebookresearch/spot-sim2real/issues/124.

This comment will be resolved as a result of this issue.

KavitShah1998 replied:
PR -> https://github.com/facebookresearch/spot-sim2real/pull/127

zephirefaith commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
-2. `vrs-file` (Input: string): This is the name of the VRS file. Ignore the `.vrs` in the name of the file. For eg. if the name of VRS file is `sample_trajectory.vrs`, then the input will be like `--vrs-file="sample_trajectory"`. This file will be searched for within the directory specified with `--data-path`
-3. `--dry-run` (Input: bool): This is used to indicate if it should command robot to move or not. If dry run is true, then it will not send the commands to the robot to make it move. *But this will still require you to have robot watching the QR code for localizing itself in the system*.
-4. `--verbose` (Input: bool): This flag is for updating the visual stream with detections from both QR & object detectors.
+### TF Frames used for Spot-Aria integration
```
@KavitShah1998 could you take care of this. thanks.

KavitShah1998 replied:
@rutadesai  @zephirefaith  I have added the information about libraries inside the header "More Information" here - https://github.com/facebookresearch/spot-sim2real/tree/aria_streamer_live?tab=readme-ov-file#more-information

Can you please verify if it makes sense and looks good?

rutadesai replied:
this was great! thanks for adding this info, @KavitShah1998 ! super useful :)

zephirefaith commented on aria_data_loaders/aria_data_utils/aria_streamer.py:
Created an issue for this and not addressing in current PR to prioritize merge + iterative changes.

KavitShah1998 commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
+- aria:       Aria's cpf frame (defined by location of Aria's camera). This frame is published by AriaLiveReader node everytime it receives a pose from Aria.
+
+Extra Aria frames (for information only):
+- cpf: Center frame of Aria with Z pointing out, X pointing up and Y pointing left
```
"aria" is the frame name we are using in P1 demo context
"cpf" is the frame name within Aria's APIs.

This is similar to "spot" vs "body" frame for Spot
"spot" is the frame used in P1 demo context
"body" is the frame name with Spot's APIs.

@akshararai does it make sense? @zephirefaith has also added a diagram for visualization of CPF and other Aria internal frames.

KavitShah1998 commented on README.md:
```diff
@@ -182,19 +182,110 @@ git checkout main
+- spot:       Spot's base frame (defined by current location of spot). This frame is published by Proprioception node in local_launch shell script
+- marker:     Marker frame (defined by the center of spot dock's april tag). This frame is published by Episodic_memory_robotic_fetch node at the start of the node when spot looks at the marker and detects it
+- ariaWorld:  Aria's world frame (defined by location where Aria's SLAM pipeline runs). This frame is published by AriaLiveReader node
+- aria:       Aria's cpf frame (defined by location of Aria's camera). This frame is published by AriaLiveReader node everytime it receives a pose from Aria.
```
You can visualize the CPF frame of aria in this figure : https://github.com/facebookresearch/spot-sim2real/blob/aria_streamer_live/aria_data_loaders/aria_frames_info.png

KavitShah1998 commented on aria_data_loaders/aria_data_utils/episodic_memory_robotic_fetch.py:
```diff
@@ -0,0 +1,467 @@
+
+# For information on useful reference frame, please check https://github.com/facebookresearch/spot-sim2real?tab=readme-ov-file#eyeglasses-run-spot-aria-project-code
+
+SPOT_DOCK_OBSERVER_WAYPOINT_LEFT = [
```
Just added a comment about them.

Basically, instead of starting spot in a pose where it can see the QR code, we are now starting it on the dock itself.

At the start of the demo, robot gets up from the dock, moves to the `SPOT_DOCK_OBSERVER_WAYPOINT_LEFT`, sits down there and starts looking at the QR code to register the marker in the TF tree.

For this specific purpose, we have 2 waypoints `SPOT_DOCK_OBSERVER_WAYPOINT_LEFT` and `SPOT_DOCK_OBSERVER_WAYPOINT_RIGHT`. User can choose between the 2 based on his layout of space.

rutadesai commented on aria_data_loaders/aria_data_utils/episodic_memory_robotic_fetch.py:
```diff
@@ -0,0 +1,474 @@
+
+    def __init__(self, spot: Spot, verbose: bool = True, use_policies: bool = True):
+        # Get Spot Skill Manager
+        # @FIXME: This is initializing a ros-node silently (at core of inheritence in
```
should we add this to our next PR, where we clean up functions and move them to utils etc?

KavitShah1998 replied:
For this particular FIXME, we have an issue up (https://github.com/facebookresearch/spot-sim2real/issues/115).
This one in particular is more involved issue and might need deeper changes in code.

But I agree about other cleanups. We have started working on the migration and cleanup of utils to a separate package in spot-sim2real @rutadesai

rutadesai reviewed (approved):
WOOHOO! Looking good :) Let's get this landed!

KavitShah1998 commented on aria_data_loaders/aria_data_utils/rviz/spot-aria.rviz:
Thank you @zephirefaith for taking care of this one.

KavitShah1998 commented:
All major comments on the PR have been addressed.

Un-addressed comments have been resolved and marked in the issue - https://github.com/facebookresearch/spot-sim2real/issues/129.

Merging this PR (https://github.com/facebookresearch/spot-sim2real/pull/110) into main.

Thank you all!
